### ISSUE TITLE
wait_for_node problems

### ISSUE THREAD

**Issue by davmlaw** [SACGF/variantgrid#346]
When we wish to wait on an existing node, we use "wait_for_node" which basically sleeps/polls until the node is ready.

The point of this is to be able to add an existing node load into our celery task dependency graph.

But if the node never finishes (eg celery job was killed and not restarted) or the total number of wait_for_node uses up all the celery worker queues, then nothing will ever get done.

Maybe instead of constructing a dependency graph, we could have another worker adding them to the queue?

**Comment by davmlaw** [comment]
Downgrading from high priority as same issue occurs in VG2

**Comment by davmlaw** [comment]
Another issue is that if this times out, it makes an event log event, but should really set the node to ERROR

Otherwise it looks like its still working

**Comment by davmlaw** [comment]
Same issue exists in wait_for_cache_task - see https://rollbar.com/jimmy.andrews/VariantGrid/items/2614/?item_page=0&#instances

Basically:

* Sometimes a node or cache can be loading, and we attach a new dependency below it
* We need to run new jobs that depend on the already loading node or cache
* There is no way to build a celery dependency for an already running task
* We thus put a "wait for node / cache" task in place, which polls/sleeps to hold up the rest of the dependencies etc

The thing is, any jobs you schedule are just going to wait there. So maybe its best to wait for the already running dependencies to complete, and then once done, build a dependency with all the tasks (that run not wait)

* Make a new state - waiting to be scheduled? Or is "dirty" already ok and will load loading screen?
* We could attach something signifying a dependency was attached to a running node (benchmark how fast it is to see if any descendants are waiting on you - may not be worth it)
* Upon finishing node tasks (or cache tasks) each node looks to see if need to schedule children to execute (do in single workers node)
* To deal with race condition where node we're waiting on finishes before we setup the "schedule children" stuff, we should run another job with a small delay to guarantee something executes and performs the scheduling. Since it runs through the single worker thread we can exit here if already done

**Comment by davmlaw** [comment]
Some good steps to recreate this issue here: https://github.com/SACGF/variantgrid_sapath/issues/175

**Comment by davmlaw** [comment]
🤖 Written by Claude

Filed #1553 as a follow-up. Once the new event-driven scheduler from this issue lands, `_add_jobs_for_group` becomes a natural choke-point for fusing per-layer node loads into single CTE-shared SQL — biggest payoff on bulk-load paths (create-from-template, force reload). Steady-state single-node behaviour is unchanged.

Contingent on this issue: the seam doesn't exist in the current scheduler.

**Comment by davmlaw** [comment]
🤖 Written by Claude

Updated `claude/analysis_node_scheduling_plan.md` to a state-driven **leased** scheduler (modelled on the `mocha/actions` reference implementation):

- Central `select_for_update(skip_locked=True)` leasing on `scheduling_single_worker` + one-task-per-node fan-out, replacing `wait_for_node`/chains/groups. A worker only ever runs one node whose dependencies are already satisfied, then re-triggers the dispatcher — never blocking on a parent.
- Lease-expiry reclamation + a periodic safety-net sweep, so a killed/lost worker self-heals instead of stalling the analysis forever (and exhausted nodes get marked ERROR so children cascade to ERROR_WITH_PARENT rather than appearing to still run).
- Exponential backoff (`run_after`) on transient `OperationalError`s, bounded by max attempts.
- All claiming/assignment funnels through the single `scheduling_single_worker`; `analysis_workers` only write the outcome of the node they already hold.

Acceptance bar remains the two `@expectedFailure` tests in `analysis/tests/test_scheduler.py`.

**Comment by davmlaw** [comment]
Testing vg test - passed
